File: modules/analytics/mobile/lib/core/services/error_tracking_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';

/// Severity level for error tracking
enum SeverityLevel {
  fatal,
  error,
  warning,
  info,
  debug,
}

/// Error tracking configuration
class ErrorTrackingConfig {
  /// Sentry DSN (Data Source Name)
  final String dsn;

  /// Environment name (production, staging, development)
  final String environment;

  /// Release/version identifier
  final String? release;

  /// Distribution identifier (e.g., app store, internal)
  final String? dist;

  /// Sample rate for transactions (0.0 to 1.0)
  final double tracesSampleRate;

  /// Enable debug mode
  final bool debug;

  /// Attach screenshots on errors
  final bool attachScreenshot;

  /// Attach view hierarchy on errors
  final bool attachViewHierarchy;

  /// Maximum number of breadcrumbs to keep
  final int maxBreadcrumbs;

  /// Enable automatic session tracking
  final bool enableAutoSessionTracking;

  /// ANR (Application Not Responding) timeout in milliseconds
  final int anrTimeoutMs;

  /// Tags to add to all events
  final Map<String, String>? tags;

  const ErrorTrackingConfig({
    required this.dsn,
    this.environment = 'production',
    this.release,
    this.dist,
    this.tracesSampleRate = 0.1,
    this.debug = false,
    this.attachScreenshot = true,
    this.attachViewHierarchy = true,
    this.maxBreadcrumbs = 100,
    this.enableAutoSessionTracking = true,
    this.anrTimeoutMs = 5000,
    this.tags,
  });
}

/// Error context for additional information
class ErrorContext {
  /// User information
  final UserContext? user;

  /// Additional tags
  final Map<String, String>? tags;

  /// Extra context data
  final Map<String, dynamic>? extra;

  /// Fingerprint for grouping
  final List<String>? fingerprint;

  /// Error level override
  final SeverityLevel? level;

  const ErrorContext({
    this.user,
    this.tags,
    this.extra,
    this.fingerprint,
    this.level,
  });
}

/// User context for error tracking
class UserContext {
  final String? id;
  final String? email;
  final String? username;
  final String? ipAddress;
  final Map<String, String>? extra;

  const UserContext({
    this.id,
    this.email,
    this.username,
    this.ipAddress,
    this.extra,
  });

  Map<String, String?> toMap() {
    return {
      if (id != null) 'id': id,
      if (email != null) 'email': email,
      if (username != null) 'username': username,
      if (ipAddress != null) 'ip_address': ipAddress,
      ...?extra,
    };
  }
}

/// Breadcrumb for debugging
class Breadcrumb {
  /// Category (e.g., 'http', 'navigation', 'user')
  final String? category;

  /// Breadcrumb message
  final String? message;

  /// Breadcrumb level
  final SeverityLevel level;

  /// Type (e.g., 'http', 'navigation', 'default')
  final String type;

  /// Additional data
  final Map<String, dynamic>? data;

  /// Timestamp
  final DateTime timestamp;

  Breadcrumb({
    this.category,
    this.message,
    this.level = SeverityLevel.info,
    this.type = 'default',
    this.data,
    DateTime? timestamp,
  }) : timestamp = timestamp ?? DateTime.now();
}

/// Abstract error tracking provider interface
abstract class ErrorTrackingProviderInterface {
  Future<void> init(ErrorTrackingConfig config);
  Future<String?> captureException(dynamic exception, {StackTrace? stackTrace, ErrorContext? context});
  Future<String?> captureMessage(String message, {SeverityLevel level, ErrorContext? context});
  void setUser(UserContext? user);
  void addBreadcrumb(Breadcrumb breadcrumb);
  void setTag(String key, String value);
  void setTags(Map<String, String> tags);
  void setExtra(String key, dynamic value);
  void setExtras(Map<String, dynamic> extras);
  void setContext(String name, Map<String, dynamic> context);
  Future<void> flush({Duration timeout = const Duration(seconds: 2)});
  Future<void> close();
}

/// Sentry provider implementation
class SentryProvider implements ErrorTrackingProviderInterface {
  // Note: Sentry instance is managed through sentry_flutter package
  // import 'package:sentry_flutter/sentry_flutter.dart';

  bool _initialized = false;
  bool _debug = false;
  ErrorTrackingConfig? _config;

  @override
  Future<void> init(ErrorTrackingConfig config) async {
    _config = config;
    _debug = config.debug;

    if (config.dsn.isEmpty) {
      _log('Sentry DSN not provided, error tracking disabled');
      return;
    }

    try {
      // Uncomment when sentry_flutter is installed:
      // await SentryFlutter.init(
      //   (options) {
      //     options.dsn = config.dsn;
      //     options.environment = config.environment;
      //     options.release = config.release;
      //     options.dist = config.dist;
      //     options.tracesSampleRate = config.tracesSampleRate;
      //     options.debug = config.debug;
      //     options.attachScreenshot = config.attachScreenshot;
      //     options.attachViewHierarchy = config.attachViewHierarchy;
      //     options.maxBreadcrumbs = config.maxBreadcrumbs;
      //     options.enableAutoSessionTracking = config.enableAutoSessionTracking;
      //     options.anrTimeoutInterval = Duration(milliseconds: config.anrTimeoutMs);
      //   },
      // );

      // Set initial tags
      // if (config.tags != null) {
      //   config.tags!.forEach((key, value) {
      //     Sentry.configureScope((scope) => scope.setTag(key, value));
      //   });
      // }

      _initialized = true;
      _log('Sentry initialized successfully');
    } catch (e) {
      debugPrint('[ErrorTracking] Failed to initialize Sentry: $e');
      debugPrint('[ErrorTracking] Install sentry_flutter for error tracking');
    }
  }

  @override
  Future<String?> captureException(
    dynamic exception, {
    StackTrace? stackTrace,
    ErrorContext? context,
  }) async {
    if (!_initialized) {
      debugPrint('[ErrorTracking] Exception (Sentry not initialized): $exception');
      if (stackTrace != null) {
        debugPrint(stackTrace.toString());
      }
      return null;
    }

    try {
      // Uncomment when sentry_flutter is installed:
      // final sentryId = await Sentry.captureException(
      //   exception,
      //   stackTrace: stackTrace,
      //   withScope: (scope) {
      //     if (context != null) {
      //       _applyContext(scope, context);
      //     }
      //   },
      // );
      // _log('Captured exception: ${sentryId.toString()}');
      // return sentryId.toString();

      _log('Would capture exception: $exception');
      return 'mock_event_id';
    } catch (e) {
      _log('Failed to capture exception: $e');
      return null;
    }
  }

  @override
  Future<String?> captureMessage(
    String message, {
    SeverityLevel level = SeverityLevel.info,
    ErrorContext? context,
  }) async {
    if (!_initialized) {
      debugPrint('[ErrorTracking:${level.name}] $message');
      return null;
    }

    try {
      // Uncomment when sentry_flutter is installed:
      // final sentryLevel = _mapSeverityLevel(level);
      // final sentryId = await Sentry.captureMessage(
      //   message,
      //   level: sentryLevel,
      //   withScope: (scope) {
      //     if (context != null) {
      //       _applyContext(scope, context);
      //     }
      //   },
      // );
      // _log('Captured message: ${sentryId.toString()}');
      // return sentryId.toString();

      _log('Would capture message: $message');
      return 'mock_event_id';
    } catch (e) {
      _log('Failed to capture message: $e');
      return null;
    }
  }

  @override
  void setUser(UserContext? user) {
    if (!_initialized) return;

    try {
      // Uncomment when sentry_flutter is installed:
      // if (user != null) {
      //   Sentry.configureScope((scope) {
      //     scope.setUser(SentryUser(
      //       id: user.id,
      //       email: user.email,
      //       username: user.username,
      //       ipAddress: user.ipAddress,
      //       data: user.extra,
      //     ));
      //   });
      // } else {
      //   Sentry.configureScope((scope) => scope.setUser(null));
      // }

      _log('Set user: ${user?.id}');
    } catch (e) {
      _log('Failed to set user: $e');
    }
  }

  @override
  void addBreadcrumb(Breadcrumb breadcrumb) {
    if (!_initialized) return;

    try {
      // Uncomment when sentry_flutter is installed:
      // Sentry.addBreadcrumb(SentryBreadcrumb(
      //   category: breadcrumb.category,
      //   message: breadcrumb.message,
      //   level: _mapSeverityLevel(breadcrumb.level),
      //   type: breadcrumb.type,
      //   data: breadcrumb.data,
      //   timestamp: breadcrumb.timestamp,
      // ));

      _log('Added breadcrumb: ${breadcrumb.message}');
    } catch (e) {
      _log('Failed to add breadcrumb: $e');
    }
  }

  @override
  void setTag(String key, String value) {
    if (!_initialized) return;

    // Uncomment when sentry_flutter is installed:
    // Sentry.configureScope((scope) => scope.setTag(key, value));
  }

  @override
  void setTags(Map<String, String> tags) {
    if (!_initialized) return;

    // Uncomment when sentry_flutter is installed:
    // Sentry.configureScope((scope) {
    //   tags.forEach((key, value) => scope.setTag(key, value));
    // });
  }

  @override
  void setExtra(String key, dynamic value) {
    if (!_initialized) return;

    // Uncomment when sentry_flutter is installed:
    // Sentry.configureScope((scope) => scope.setExtra(key, value));
  }

  @override
  void setExtras(Map<String, dynamic> extras) {
    if (!_initialized) return;

    // Uncomment when sentry_flutter is installed:
    // Sentry.configureScope((scope) {
    //   extras.forEach((key, value) => scope.setExtra(key, value));
    // });
  }

  @override
  void setContext(String name, Map<String, dynamic> context) {
    if (!_initialized) return;

    // Uncomment when sentry_flutter is installed:
    // Sentry.configureScope((scope) => scope.setContexts(name, context));
  }

  @override
  Future<void> flush({Duration timeout = const Duration(seconds: 2)}) async {
    if (!_initialized) return;

    // Uncomment when sentry_flutter is installed:
    // await Sentry.flush(timeout: timeout);
  }

  @override
  Future<void> close() async {
    if (!_initialized) return;

    // Uncomment when sentry_flutter is installed:
    // await Sentry.close();
  }

  // Helper to map severity levels
  // SentryLevel _mapSeverityLevel(SeverityLevel level) {
  //   switch (level) {
  //     case SeverityLevel.fatal:
  //       return SentryLevel.fatal;
  //     case SeverityLevel.error:
  //       return SentryLevel.error;
  //     case SeverityLevel.warning:
  //       return SentryLevel.warning;
  //     case SeverityLevel.info:
  //       return SentryLevel.info;
  //     case SeverityLevel.debug:
  //       return SentryLevel.debug;
  //   }
  // }

  // Helper to apply context to scope
  // void _applyContext(Scope scope, ErrorContext context) {
  //   if (context.user != null) {
  //     scope.setUser(SentryUser(
  //       id: context.user!.id,
  //       email: context.user!.email,
  //       username: context.user!.username,
  //     ));
  //   }
  //   if (context.tags != null) {
  //     context.tags!.forEach((key, value) => scope.setTag(key, value));
  //   }
  //   if (context.extra != null) {
  //     context.extra!.forEach((key, value) => scope.setExtra(key, value));
  //   }
  //   if (context.fingerprint != null) {
  //     scope.fingerprint = context.fingerprint!;
  //   }
  //   if (context.level != null) {
  //     scope.level = _mapSeverityLevel(context.level!);
  //   }
  // }

  void _log(String message) {
    if (_debug) {
      debugPrint('[ErrorTracking:Sentry] $message');
    }
  }
}

/// No-op provider (fallback when Sentry is not configured)
class NoopProvider implements ErrorTrackingProviderInterface {
  @override
  Future<void> init(ErrorTrackingConfig config) async {}

  @override
  Future<String?> captureException(
    dynamic exception, {
    StackTrace? stackTrace,
    ErrorContext? context,
  }) async {
    debugPrint('[ErrorTracking:Noop] Exception: $exception');
    if (stackTrace != null) {
      debugPrint(stackTrace.toString());
    }
    return null;
  }

  @override
  Future<String?> captureMessage(
    String message, {
    SeverityLevel level = SeverityLevel.info,
    ErrorContext? context,
  }) async {
    debugPrint('[ErrorTracking:Noop:${level.name}] $message');
    return null;
  }

  @override
  void setUser(UserContext? user) {}

  @override
  void addBreadcrumb(Breadcrumb breadcrumb) {}

  @override
  void setTag(String key, String value) {}

  @override
  void setTags(Map<String, String> tags) {}

  @override
  void setExtra(String key, dynamic value) {}

  @override
  void setExtras(Map<String, dynamic> extras) {}

  @override
  void setContext(String name, Map<String, dynamic> context) {}

  @override
  Future<void> flush({Duration timeout = const Duration(seconds: 2)}) async {}

  @override
  Future<void> close() async {}
}

/// Main Error Tracking Service
class ErrorTrackingService {
  static final ErrorTrackingService _instance = ErrorTrackingService._internal();

  factory ErrorTrackingService() => _instance;

  ErrorTrackingService._internal();

  ErrorTrackingProviderInterface _provider = NoopProvider();
  bool _initialized = false;

  /// Check if error tracking is initialized
  bool get isInitialized => _initialized;

  /// Initialize error tracking with Sentry
  ///
  /// Call this in your main.dart before runApp:
  /// ```dart
  /// void main() async {
  ///   WidgetsFlutterBinding.ensureInitialized();
  ///
  ///   await errorTracking.init(ErrorTrackingConfig(
  ///     dsn: 'your-sentry-dsn',
  ///     environment: kReleaseMode ? 'production' : 'development',
  ///   ));
  ///
  ///   runApp(MyApp());
  /// }
  /// ```
  Future<void> init(ErrorTrackingConfig config) async {
    if (_initialized) {
      debugPrint('[ErrorTracking] Already initialized');
      return;
    }

    if (config.dsn.isEmpty) {
      _provider = NoopProvider();
      _initialized = true;
      return;
    }

    _provider = SentryProvider();
    await _provider.init(config);
    _initialized = true;

    // Set up Flutter error handlers
    _setupFlutterErrorHandlers();
  }

  /// Set up Flutter error handlers to capture all errors
  void _setupFlutterErrorHandlers() {
    // Capture Flutter framework errors
    FlutterError.onError = (FlutterErrorDetails details) {
      captureException(
        details.exception,
        stackTrace: details.stack,
        context: ErrorContext(
          tags: {'error.type': 'flutter_error'},
          extra: {
            'library': details.library,
            'context': details.context?.toString(),
            'informationCollector': details.informationCollector?.call().map((e) => e.toString()).toList(),
          },
        ),
      );

      // Also print to console in debug mode
      if (kDebugMode) {
        FlutterError.dumpErrorToConsole(details);
      }
    };

    // Capture errors that escape the Flutter framework
    PlatformDispatcher.instance.onError = (error, stack) {
      captureException(
        error,
        stackTrace: stack,
        context: const ErrorContext(
          tags: {'error.type': 'platform_error'},
        ),
      );
      return true;
    };
  }

  /// Capture an exception and send to Sentry
  Future<String?> captureException(
    dynamic exception, {
    StackTrace? stackTrace,
    ErrorContext? context,
  }) {
    return _provider.captureException(exception, stackTrace: stackTrace, context: context);
  }

  /// Capture a message and send to Sentry
  Future<String?> captureMessage(
    String message, {
    SeverityLevel level = SeverityLevel.info,
    ErrorContext? context,
  }) {
    return _provider.captureMessage(message, level: level, context: context);
  }

  /// Set user context for all future events
  void setUser(UserContext? user) {
    _provider.setUser(user);
  }

  /// Add a breadcrumb for debugging
  void addBreadcrumb(Breadcrumb breadcrumb) {
    _provider.addBreadcrumb(breadcrumb);
  }

  /// Set a tag for all future events
  void setTag(String key, String value) {
    _provider.setTag(key, value);
  }

  /// Set multiple tags for all future events
  void setTags(Map<String, String> tags) {
    _provider.setTags(tags);
  }

  /// Set extra data for all future events
  void setExtra(String key, dynamic value) {
    _provider.setExtra(key, value);
  }

  /// Set multiple extra data for all future events
  void setExtras(Map<String, dynamic> extras) {
    _provider.setExtras(extras);
  }

  /// Set a named context for all future events
  void setContext(String name, Map<String, dynamic> context) {
    _provider.setContext(name, context);
  }

  /// Flush pending events (call before app shutdown)
  Future<void> flush({Duration timeout = const Duration(seconds: 2)}) {
    return _provider.flush(timeout: timeout);
  }

  /// Close the SDK
  Future<void> close() {
    return _provider.close();
  }

  /// Wrap a function with error tracking
  ///
  /// Usage:
  /// ```dart
  /// await errorTracking.wrap(() async {
  ///   // Your code here
  /// });
  /// ```
  Future<T> wrap<T>(Future<T> Function() fn, {ErrorContext? context}) async {
    try {
      return await fn();
    } catch (e, stack) {
      await captureException(e, stackTrace: stack, context: context);
      rethrow;
    }
  }

  /// Wrap a synchronous function with error tracking
  T wrapSync<T>(T Function() fn, {ErrorContext? context}) {
    try {
      return fn();
    } catch (e, stack) {
      captureException(e, stackTrace: stack, context: context);
      rethrow;
    }
  }
}

/// Global error tracking instance
final errorTracking = ErrorTrackingService();

/// Helper function to run app with Sentry initialization
///
/// Usage in main.dart:
/// ```dart
/// void main() async {
///   await runAppWithErrorTracking(
///     config: ErrorTrackingConfig(
///       dsn: 'your-sentry-dsn',
///       environment: kReleaseMode ? 'production' : 'development',
///     ),
///     appRunner: () => runApp(MyApp()),
///   );
/// }
/// ```
Future<void> runAppWithErrorTracking({
  required ErrorTrackingConfig config,
  required void Function() appRunner,
}) async {
  await runZonedGuarded(
    () async {
      WidgetsFlutterBinding.ensureInitialized();
      await errorTracking.init(config);
      appRunner();
    },
    (error, stackTrace) {
      errorTracking.captureException(
        error,
        stackTrace: stackTrace,
        context: const ErrorContext(
          tags: {'error.type': 'zone_error'},
        ),
      );
    },
  );
}

/// Navigation observer for automatic screen breadcrumbs
class ErrorTrackingNavigatorObserver extends NavigatorObserver {
  @override
  void didPush(Route<dynamic> route, Route<dynamic>? previousRoute) {
    super.didPush(route, previousRoute);
    _trackNavigation(route, 'push');
  }

  @override
  void didPop(Route<dynamic> route, Route<dynamic>? previousRoute) {
    super.didPop(route, previousRoute);
    if (previousRoute != null) {
      _trackNavigation(previousRoute, 'pop');
    }
  }

  @override
  void didReplace({Route<dynamic>? newRoute, Route<dynamic>? oldRoute}) {
    super.didReplace(newRoute: newRoute, oldRoute: oldRoute);
    if (newRoute != null) {
      _trackNavigation(newRoute, 'replace');
    }
  }

  void _trackNavigation(Route<dynamic> route, String action) {
    final screenName = route.settings.name;
    if (screenName != null && screenName.isNotEmpty) {
      errorTracking.addBreadcrumb(Breadcrumb(
        category: 'navigation',
        message: '$action: $screenName',
        type: 'navigation',
        data: {
          'screen': screenName,
          'action': action,
          'arguments': route.settings.arguments?.toString(),
        },
      ));
    }
  }
}

File: modules/analytics/backend/src/services/error-tracking.service.ts
/**
 * Error Tracking Service (Sentry Integration)
 *
 * Premium tier feature for comprehensive error tracking and performance monitoring.
 *
 * Usage:
 *   import { errorTracking } from './services/error-tracking.service';
 *   errorTracking.initErrorTracking({ dsn: 'your-sentry-dsn' });
 *   errorTracking.captureException(error, { userId: 'user_123' });
 */

// =============================================================================
// Types
// =============================================================================

export type SeverityLevel = 'fatal' | 'error' | 'warning' | 'info' | 'debug' | 'log';

export interface ErrorTrackingConfig {
  /** Sentry DSN (Data Source Name) */
  dsn: string;
  /** Environment name (production, staging, development) */
  environment?: string;
  /** Release/version identifier */
  release?: string;
  /** Sample rate for transactions (0.0 to 1.0) */
  tracesSampleRate?: number;
  /** Sample rate for profiles (0.0 to 1.0) */
  profilesSampleRate?: number;
  /** Enable debug mode */
  debug?: boolean;
  /** Server name identifier */
  serverName?: string;
  /** Tags to add to all events */
  tags?: Record<string, string>;
  /** Enable performance monitoring */
  enableTracing?: boolean;
  /** Integrations to ignore (e.g., specific routes) */
  ignoreTransactions?: string[];
}

export interface ErrorContext {
  /** User information */
  user?: {
    id?: string;
    email?: string;
    username?: string;
    [key: string]: string | undefined;
  };
  /** Additional tags */
  tags?: Record<string, string>;
  /** Extra context data */
  extra?: Record<string, unknown>;
  /** Fingerprint for grouping */
  fingerprint?: string[];
  /** Error level override */
  level?: SeverityLevel;
}

export interface Breadcrumb {
  /** Category (e.g., 'http', 'navigation', 'user') */
  category?: string;
  /** Breadcrumb message */
  message?: string;
  /** Breadcrumb level */
  level?: SeverityLevel;
  /** Type (e.g., 'http', 'navigation', 'default') */
  type?: string;
  /** Additional data */
  data?: Record<string, unknown>;
  /** Timestamp (defaults to now) */
  timestamp?: number;
}

export interface UserContext {
  id?: string;
  email?: string;
  username?: string;
  ip_address?: string;
  [key: string]: string | undefined;
}

export interface TransactionContext {
  /** Transaction name */
  name: string;
  /** Operation type (e.g., 'http.server', 'db.query') */
  op: string;
  /** Transaction description */
  description?: string;
  /** Transaction tags */
  tags?: Record<string, string>;
  /** Transaction data */
  data?: Record<string, unknown>;
}

// =============================================================================
// Error Tracking Provider Interface
// =============================================================================

interface ErrorTrackingProviderInterface {
  init(config: ErrorTrackingConfig): void;
  captureException(error: Error | unknown, context?: ErrorContext): string | undefined;
  captureMessage(message: string, level?: SeverityLevel, context?: ErrorContext): string | undefined;
  setUser(user: UserContext | null): void;
  addBreadcrumb(breadcrumb: Breadcrumb): void;
  setTag(key: string, value: string): void;
  setTags(tags: Record<string, string>): void;
  setExtra(key: string, value: unknown): void;
  setExtras(extras: Record<string, unknown>): void;
  setContext(name: string, context: Record<string, unknown>): void;
  startTransaction(context: TransactionContext): Transaction | null;
  flush(timeout?: number): Promise<boolean>;
  close(timeout?: number): Promise<boolean>;
}

// =============================================================================
// Transaction Interface
// =============================================================================

export interface Transaction {
  name: string;
  op: string;
  setTag(key: string, value: string): void;
  setData(key: string, value: unknown): void;
  startChild(context: { op: string; description?: string }): Span;
  finish(): void;
}

export interface Span {
  setTag(key: string, value: string): void;
  setData(key: string, value: unknown): void;
  finish(): void;
}

// =============================================================================
// Sentry Provider Implementation
// =============================================================================

class SentryProvider implements ErrorTrackingProviderInterface {
  private sentry: typeof import('@sentry/node') | null = null;
  private config: ErrorTrackingConfig | null = null;
  private initialized = false;
  private debug = false;

  init(config: ErrorTrackingConfig): void {
    this.config = config;
    this.debug = config.debug || false;

    if (!config.dsn) {
      this.log('Sentry DSN not provided, error tracking disabled');
      return;
    }

    try {
      // Dynamic import to allow graceful fallback if @sentry/node is not installed
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const Sentry = require('@sentry/node');
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const { nodeProfilingIntegration } = require('@sentry/profiling-node');

      Sentry.init({
        dsn: config.dsn,
        environment: config.environment || process.env.NODE_ENV || 'development',
        release: config.release || process.env.npm_package_version,
        debug: config.debug,
        serverName: config.serverName,
        tracesSampleRate: config.tracesSampleRate ?? 0.1,
        profilesSampleRate: config.profilesSampleRate ?? 0.1,
        integrations: config.enableTracing !== false ? [nodeProfilingIntegration()] : [],
        beforeSend: (event) => {
          // Filter out certain errors if needed
          return event;
        },
        beforeSendTransaction: (event) => {
          // Filter out certain transactions if needed
          if (config.ignoreTransactions?.includes(event.transaction || '')) {
            return null;
          }
          return event;
        },
      });

      // Set initial tags
      if (config.tags) {
        Object.entries(config.tags).forEach(([key, value]) => {
          Sentry.setTag(key, value);
        });
      }

      this.sentry = Sentry;
      this.initialized = true;
      this.log('Sentry initialized successfully');
    } catch (error) {
      console.warn('[ErrorTracking] Failed to initialize Sentry:', error);
      console.warn('[ErrorTracking] Install @sentry/node and @sentry/profiling-node for error tracking');
    }
  }

  captureException(error: Error | unknown, context?: ErrorContext): string | undefined {
    if (!this.sentry || !this.initialized) {
      this.logError(error);
      return undefined;
    }

    return this.sentry.withScope((scope) => {
      if (context) {
        this.applyContext(scope, context);
      }

      const eventId = this.sentry!.captureException(error);
      this.log('Captured exception:', eventId);
      return eventId;
    });
  }

  captureMessage(message: string, level: SeverityLevel = 'info', context?: ErrorContext): string | undefined {
    if (!this.sentry || !this.initialized) {
      console.log(`[ErrorTracking:${level}]`, message);
      return undefined;
    }

    return this.sentry.withScope((scope) => {
      if (context) {
        this.applyContext(scope, context);
      }

      const eventId = this.sentry!.captureMessage(message, level);
      this.log('Captured message:', eventId);
      return eventId;
    });
  }

  setUser(user: UserContext | null): void {
    if (!this.sentry || !this.initialized) return;
    this.sentry.setUser(user);
    this.log('Set user:', user?.id);
  }

  addBreadcrumb(breadcrumb: Breadcrumb): void {
    if (!this.sentry || !this.initialized) return;
    this.sentry.addBreadcrumb({
      ...breadcrumb,
      timestamp: breadcrumb.timestamp || Date.now() / 1000,
    });
  }

  setTag(key: string, value: string): void {
    if (!this.sentry || !this.initialized) return;
    this.sentry.setTag(key, value);
  }

  setTags(tags: Record<string, string>): void {
    if (!this.sentry || !this.initialized) return;
    this.sentry.setTags(tags);
  }

  setExtra(key: string, value: unknown): void {
    if (!this.sentry || !this.initialized) return;
    this.sentry.setExtra(key, value);
  }

  setExtras(extras: Record<string, unknown>): void {
    if (!this.sentry || !this.initialized) return;
    this.sentry.setExtras(extras);
  }

  setContext(name: string, context: Record<string, unknown>): void {
    if (!this.sentry || !this.initialized) return;
    this.sentry.setContext(name, context);
  }

  startTransaction(context: TransactionContext): Transaction | null {
    if (!this.sentry || !this.initialized) return null;

    const transaction = this.sentry.startSpan({
      name: context.name,
      op: context.op,
    });

    if (!transaction) return null;

    // Wrap in our Transaction interface
    return {
      name: context.name,
      op: context.op,
      setTag: (key: string, value: string) => transaction.setAttribute(key, value),
      setData: (key: string, value: unknown) => transaction.setAttribute(key, String(value)),
      startChild: (childContext: { op: string; description?: string }) => {
        const span = this.sentry!.startSpan({
          name: childContext.description || childContext.op,
          op: childContext.op,
        });
        return {
          setTag: (key: string, value: string) => span?.setAttribute(key, value),
          setData: (key: string, value: unknown) => span?.setAttribute(key, String(value)),
          finish: () => span?.end(),
        };
      },
      finish: () => transaction.end(),
    };
  }

  async flush(timeout = 2000): Promise<boolean> {
    if (!this.sentry || !this.initialized) return true;
    return this.sentry.flush(timeout);
  }

  async close(timeout = 2000): Promise<boolean> {
    if (!this.sentry || !this.initialized) return true;
    return this.sentry.close(timeout);
  }

  private applyContext(scope: import('@sentry/node').Scope, context: ErrorContext): void {
    if (context.user) {
      scope.setUser(context.user);
    }
    if (context.tags) {
      Object.entries(context.tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }
    if (context.extra) {
      Object.entries(context.extra).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
    }
    if (context.fingerprint) {
      scope.setFingerprint(context.fingerprint);
    }
    if (context.level) {
      scope.setLevel(context.level);
    }
  }

  private log(...args: unknown[]): void {
    if (this.debug) {
      console.log('[ErrorTracking:Sentry]', ...args);
    }
  }

  private logError(error: unknown): void {
    console.error('[ErrorTracking] Error (Sentry not initialized):', error);
  }
}

// =============================================================================
// No-op Provider (fallback when Sentry is not configured)
// =============================================================================

class NoopProvider implements ErrorTrackingProviderInterface {
  init(): void {}
  captureException(error: Error | unknown): string | undefined {
    console.error('[ErrorTracking:Noop] Exception:', error);
    return undefined;
  }
  captureMessage(message: string, level?: SeverityLevel): string | undefined {
    console.log(`[ErrorTracking:Noop:${level || 'info'}]`, message);
    return undefined;
  }
  setUser(): void {}
  addBreadcrumb(): void {}
  setTag(): void {}
  setTags(): void {}
  setExtra(): void {}
  setExtras(): void {}
  setContext(): void {}
  startTransaction(): Transaction | null {
    return null;
  }
  async flush(): Promise<boolean> {
    return true;
  }
  async close(): Promise<boolean> {
    return true;
  }
}

// =============================================================================
// Error Tracking Service
// =============================================================================

class ErrorTrackingService {
  private provider: ErrorTrackingProviderInterface = new NoopProvider();
  private initialized = false;

  /**
   * Initialize error tracking with Sentry
   */
  initErrorTracking(config: ErrorTrackingConfig): void {
    if (this.initialized) {
      console.warn('[ErrorTracking] Already initialized');
      return;
    }

    const enabled = process.env.ERROR_TRACKING_ENABLED !== 'false';
    if (!enabled) {
      this.provider = new NoopProvider();
      this.initialized = true;
      return;
    }

    this.provider = new SentryProvider();
    this.provider.init(config);
    this.initialized = true;
  }

  /**
   * Capture an exception and send to Sentry
   */
  captureException(error: Error | unknown, context?: ErrorContext): string | undefined {
    return this.provider.captureException(error, context);
  }

  /**
   * Capture a message and send to Sentry
   */
  captureMessage(message: string, level: SeverityLevel = 'info', context?: ErrorContext): string | undefined {
    return this.provider.captureMessage(message, level, context);
  }

  /**
   * Set user context for all future events
   */
  setUser(user: UserContext | null): void {
    this.provider.setUser(user);
  }

  /**
   * Add a breadcrumb for debugging
   */
  addBreadcrumb(breadcrumb: Breadcrumb): void {
    this.provider.addBreadcrumb(breadcrumb);
  }

  /**
   * Set a tag for all future events
   */
  setTag(key: string, value: string): void {
    this.provider.setTag(key, value);
  }

  /**
   * Set multiple tags for all future events
   */
  setTags(tags: Record<string, string>): void {
    this.provider.setTags(tags);
  }

  /**
   * Set extra data for all future events
   */
  setExtra(key: string, value: unknown): void {
    this.provider.setExtra(key, value);
  }

  /**
   * Set multiple extra data for all future events
   */
  setExtras(extras: Record<string, unknown>): void {
    this.provider.setExtras(extras);
  }

  /**
   * Set a named context for all future events
   */
  setContext(name: string, context: Record<string, unknown>): void {
    this.provider.setContext(name, context);
  }

  /**
   * Start a new transaction for performance monitoring
   */
  startTransaction(context: TransactionContext): Transaction | null {
    return this.provider.startTransaction(context);
  }

  /**
   * Flush pending events (call before shutting down)
   */
  async flush(timeout?: number): Promise<boolean> {
    return this.provider.flush(timeout);
  }

  /**
   * Close the SDK (call when shutting down)
   */
  async close(timeout?: number): Promise<boolean> {
    return this.provider.close(timeout);
  }

  /**
   * Check if error tracking is initialized
   */
  get isInitialized(): boolean {
    return this.initialized;
  }
}

// =============================================================================
// Singleton Export
// =============================================================================

export const errorTracking = new ErrorTrackingService();

// =============================================================================
// Request Handler Helper
// =============================================================================

/**
 * Get request context for error tracking
 */
export function getRequestContext(req: {
  method?: string;
  url?: string;
  headers?: Record<string, string | string[] | undefined>;
  ip?: string;
  user?: { id?: string; email?: string };
}): ErrorContext {
  return {
    user: req.user
      ? {
          id: req.user.id,
          email: req.user.email,
        }
      : undefined,
    tags: {
      method: req.method || 'unknown',
      url: req.url || 'unknown',
    },
    extra: {
      headers: {
        'user-agent': req.headers?.['user-agent'],
        'x-request-id': req.headers?.['x-request-id'],
        'x-forwarded-for': req.headers?.['x-forwarded-for'],
      },
      ip: req.ip,
    },
  };
}

export default errorTracking;
